specs/vm/instruction_set.md

```diff
@@ -1386,6 +1387,42 @@ Cease VM execution and revert script effects. After a revert:
 1. All [OutputContract](../protocol/tx_format.md#outputcontract) outputs will have the same `balanceRoot` and `stateRoot` as on initialization.
 1. All [OutputVariable](../protocol/tx_format.md#outputvariable) outputs will have `to`, `amount`, and `asset_id` of zero.
 
+### SMO: Send message to output
+
+|             |                                                                                     |
+|-------------|-------------------------------------------------------------------------------------|
+| Description | Send a message to recipient address `MEM[$rA, 32]` with call abi `MEM[$rA + 32, $rB]` and `$rD` coins, with output `$rC`. |
+| Operation   | ```outputmessage(MEM[$fp, 32], MEM[$rA, 32], MEM[$rA + 32, $rB], $rD, $rC);```      |
+| Syntax      | `smo $rA, $rB, $rC, $rD`                                                            |
+| Encoding    | `0x00 rA rB rC rD`                                                                  |
+| Notes       |                                                                                     |
+
+Given helper `balanceOfStart(asset_id: byte[32]) -> uint32` which returns the memory address of the remaining free balance of `asset_id`, or panics if `asset_id` has no free balance remaining.
+
+Panic if:
+
+- `$rA + 32` overflows
+- `$rA + $rB + 32` overflows
+- `$rA + 32 > VM_MAX_RAM`
+- `$rA + $rB + 32 > VM_MAX_RAM`
+- `$rC > tx.outputsCount`
+- In an external context, if `$rD > MEM[balanceOfStart(0), 8]`
+- In an internal context, if `$rD` is greater than the balance of asset ID 0 of output with contract ID `MEM[$fp, 32]`
+- `tx.outputs[$rC].type != OutputType.Message`
+- `tx.outputs[$rC].recipient != 0 || tx.outputs[$rC].sender != 0`
+
+In an external context, decrease `MEM[balanceOfStart(0), 8]` by `$rD`. In an internal context, decrease asset ID 0 balance of output with contract ID `MEM[$fp, 32]` by `$rD`. Then set:
+
+- `tx.outputs[$rC].recipient = MEM[$rA, 32]`
+- `tx.outputs[$rC].sender = MEM[$fp, 32]`
+- `tx.outputs[$rC].callABI = MEM[$rA + 32, $rB]`
+- `tx.outputs[$rC].amount = $rD`
+- `tx.outputs[$rC].nonce = ++messageOutputNonce`
```

Comment:
Is this a global `messageOutputNonce` variable? as transactions can be executed in parallel this can become a synchronization problem. I propose that nonce be left zero in instruction and in the loop, after block is executed, we transverse all transactions and set the nonce, we already need to do that loop to create merkle tree for all MessageOutputs. And globals are not something desirable, maybe we can somehow mix fuel block height and local block Nonce?

Reply:
yeah, it was imagined as a global variable. Setting the nonce later is ok, but that also means we can't calculate the messageID until this is filled in. It can be whatever as long as it's available when an app or user looks to a processed transaction to find the outputs (to get the messageID and lookup the merkle proof for eventual relaying). What exactly do you think it should be?

Reply:
Introducing two variables, `block_height` u32 and `nonce` (that is local to the block) u32. It will break the symmetry that we have but will remove the need for global vars. MessageId needs to be unique and height/local_nonce do that. wdyt?

Reply:
So this would help remove the need for a global variable, but does it solve the problem of running txs in parallel? What if we treat the nonce like `hash(tx ++ OutputMessageIndex)` where `tx` is some sort of unique tx identifier and `OutputMessageIndex` is the index of the output message in the list of the transactions output?

Reply:
a solution that would allow running txs in parallel is to not set nonce in runtime in `SMO` but in post-processing when block is already created.
`hash(tx++OutputMessageIndex)` sound nice, it is unique and similar to [UtxoId](https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/identifiers.md#utxo-id) but this will increase our nonce size to 256bits if it is keccake hash.

Reply:
Do we actually need the nonce to be unique? Why is `txId ++ OutputMessageIndex` (no hashing!) not a sufficiently unique identifier?

Reply:
It is sufficient, it is just a matter of choosing.

Reply:
So with the nonce being expanded to 256bits now, would `txId + OutputMessageIndex` be something that can be computed at the time of execution (no need for a second pass after building a block)

Reply:
TxId+OutputIndex is good and i assume that they can be set and checked at runtime, let us go with it. 256bit for TxId and u8 for the index as MAX_OUTPUTS is 8: https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/tx_format.md#constants
A second pass is needed for merkalization of message ids. I will resolve this comments if it is okay, as consern with global nonce was mitigated.

Reply:
> would txId + OutputMessageIndex be something that can be computed at the time of execution

Actually yes! The tx ID is computed off of the non-malleable parts only (technically, the malleable parts are set to zero, but whatever). And it's actually available during VM execution at memory address zero.

Reply:
Ah so the max inputs/outputs were changed to 255 in the implementation, but never updated in the specs. And there was some discussion to change the type to `u64`, lost to the void of time.